File: python/agent_authority/checkpoint.py
"""Periodic audit-log anchoring (mirrors TS startAuditCheckpointing).

Automates the manual ``checkpoint_audit()`` / ``verify_audit_checkpoint()``
pattern: on a background thread, every ``interval_ms`` it signs the audit head
and hands the checkpoint to ``sink``. Store checkpoints **out of the audit
writer's reach** for tail-deletion / rewrite detection to hold.
"""

from __future__ import annotations

import threading
from typing import Callable


def start_audit_checkpointing(
    engine,
    *,
    interval_ms: int,
    sink: Callable[[dict], None],
    on_error: Callable[[BaseException], None] | None = None,
) -> Callable[[], None]:
    """Start checkpointing; returns a ``stop()`` function. The worker is a daemon
    thread, so it won't keep the interpreter alive on its own."""
    stop_event = threading.Event()
    interval_s = interval_ms / 1000.0

    def loop() -> None:
        while not stop_event.wait(interval_s):
            try:
                sink(engine.checkpoint_audit())
            except BaseException as e:  # noqa: BLE001
                if on_error is not None:
                    on_error(e)

    thread = threading.Thread(target=loop, daemon=True)
    thread.start()

    def stop() -> None:
        stop_event.set()
        thread.join(timeout=1.0)

    return stop

File: src/checkpoint.ts
import type { AuditCheckpoint } from "./types.js";

/** The slice of the engine this helper needs (so it doesn't import Behalf). */
export interface Checkpointable {
  checkpointAudit(): Promise<AuditCheckpoint>;
}

export interface CheckpointingOptions {
  /** How often to take a checkpoint, in ms. */
  intervalMs: number;
  /**
   * Where each checkpoint goes — store it **out of the audit writer's reach**
   * (another host, object storage, a ledger) for the guarantee to hold.
   */
  sink: (checkpoint: AuditCheckpoint) => void | Promise<void>;
  /** Called if a checkpoint or sink throws (default: ignore). */
  onError?: (err: unknown) => void;
}

/**
 * Periodically anchor the audit log: every `intervalMs`, sign the current head
 * (`engine.checkpointAudit()`) and hand it to `sink`. Returns a `stop()` function.
 * The timer is `unref`'d so it won't keep a Node process alive on its own.
 *
 * This automates the manual `checkpointAudit()` / `verifyAuditCheckpoint()`
 * pattern so tail-deletion and rewrites stay detectable without remembering to
 * checkpoint by hand — provided the sink stores checkpoints somewhere the log's
 * writer cannot reach.
 */
export function startAuditCheckpointing(
  engine: Checkpointable,
  opts: CheckpointingOptions,
): () => void {
  const tick = async (): Promise<void> => {
    try {
      await opts.sink(await engine.checkpointAudit());
    } catch (err) {
      opts.onError?.(err);
    }
  };
  const timer = setInterval(tick, opts.intervalMs);
  (timer as { unref?: () => void }).unref?.();
  return () => clearInterval(timer);
}
